doc/feature/httpServerREST.md

```diff
@@ -0,0 +1,161 @@
+    {
+        auto response = http::Response{};
+        response.contentType = "image/jpeg";
+        response.payload = doTheHeavyWork(windowId);
+        return response;
+    }
+
+    // simplified window collection for the example
+    std::set<std::string> windows{"1234", "567"};
+
+    // Register an individual endpoint for each thumbnail. Note that this is
+    // tedious to manage as the list of windows evolves. It could be more
+    // efficiently served by a generic function:
+    // server.handlePath(http::Verb::GET, "api/windows", getWindowInfo );
+    for (const auto& windowId : windows)
+    {
+        const auto makeThumbnailAsync = [windowId](const std::string&)
+        {
+            return std::async(std::launch::async, makeThumbnail, windowId);
+        };
+        server.handle(http::Verb::GET, "api/windows/"+windowId+"/thumbnail",
+                      makeThumbnailAsync);
+    }
+
+    // Here we use a generic endpoint for all window close operations. "path"
+    // provides the url part after the endpoint: "api/windows/567" -> "567"
+    const auto closeWindow = [&](const std::string& path, const std::string&)
+    {
+        if (!windows.count(path))
+            return make_ready_future(http::Response{http::Code::NOT_FOUND});
+
+        server.remove("api/windows/"+path+"/thumbnail");
+        windows.erase(path);
+        return make_ready_future(http::Response{http::Code::OK});
+    };
+    server.handlePath(http::Verb::DELETE, "api/windows", closeWindow);
+
+    // This function uses POST to open a new window and return its location.
+    const auto openWindow = [&](const std::string& request)
+    {
+        if (!windows.insert(request))
+            return make_ready_future(http::Response{http::Code::NO_CONTENT});
+
+        auto response = http::Response();
+        response.code = http::Code::CREATED;
+        response.headers[http::Header::LOCATION] = "api/windows/"+request;
+        return make_ready_future(std::move(response));
+    };
+    server.handle(http::Verb::POST, "api/windows", openWindow);
+
+## Implementation
+
+In requestHandler.h, the struct HTTPRequest needs to be changed from:
+
+    // output from zeroeq::http::Server
+    HTTPServer::connection::status_t status;
+    std::string reply;
+
+to
+
+    // output from zeroeq::http::Server
+    std::future<Response> response;
+
+In requestHandler.cpp, the ConnectionHandler must wait on response.get();
+
+In server.cpp, all existing handle*() functions must be adapted to wrap the
+response data as ready futures. Additional maps must be added for the POST,
+PATCH and DELETE actions.
+
+};
+
+## Issues
+
+### 1: How many threads should the cppnetlib server use?
+
+_Resolution: Open:_ The server thread pool consists of a single thread by
+default. Using more threads may improve performance by accepting multiple
+concurrent requests but would require changes in the zeroeq inproc signaling
+mechanism.
```

Comment:
Correct, the first idea was to add the possibility for the user to delay the response to a request without touching too much the server implementation. But what you suggest is also valid: the callback function(s) could be called directly from the cppnetlib thread(s), since it/they block on response.get() anyway. That might even make the API simpler.